### PR TITLE
Add Flax Dropout tutorial

### PR DESCRIPTION
**Pull request by 8bitmp3**
@levskaya @IvyZX @chiamp @marcvanzee

### CODE REVIEW

**Review by IvyZX — CHANGES_REQUESTED** [review]
For the `dropout_key` in `train_loop`, I believe you need to split a different key on each step so that different activation happens. So you should split `dropout_key` into two keys, use one for this step, and return the other so that the next step can take it as an input (and do more splitting). 

Other than this the doc looks good! I like the part that explains every piece of code cleanly.

**Comment by andsteing** [comment]
(removed myself as a reviewer because I don't have time to properly review this ... maybe @marcvanzee can have a look instead? )

File: docs/tutorials/dropout_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+Create a simple Flax model, subclassed from [Flax `Module`](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#module-basics). Note that:
+
+- To add a dropout layer in Flax, use [`flax.linen.Dropout`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Dropout.html#flax.linen.Dropout).
+- In `flax.linen.Dropout`, the `deterministic` argument is `None` by default. If `false`, the inputs are scaled by `1 / (1 - dropout_rate)` and masked. When it's `true`, no mask is applied (the dropout is turned off).
+
+```{code-cell}
+# A simple convolutional network with a dropout layer.
+class CNN(nn.Module):
+  training: bool
```
Comment (cgarciae) [inline docs/tutorials/dropout_tutorial.md:61]: `training`  (or `train` as its most commonly used) should be an argument to `__call__`, else you cannot modify the training state via `TrainState.apply_fn`.

File: docs/tutorials/dropout_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+- The model constructor argument should be set to `training=True`.
+
+```{code-cell}
+@jax.jit
+def train_step(state, batch):
+  """Train for a single step."""
+  def loss_fn(params):
+    # Perform the forward pass with `flax.linen.apply()`.
+    # Use the `dropout_key` for the `rngs` argument.
+    logits = CNN(training=True).apply({'params': params}, batch['image'], rngs={'dropout': dropout_key})
```
Comment (cgarciae) [inline docs/tutorials/dropout_tutorial.md:142]: If the previous is addressed this could be written as:
```python
logits = state.apply_fn(
  {'params': params}, batch['image'], train=True, rngs={'dropout': dropout_key})
```

File: docs/tutorials/dropout_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  metrics = compute_metrics(logits=logits, labels=batch['label'])
+  return state, metrics
+```
+
+Write the evaluation step function. Remember to set the model constructor argument to `training=false`.
+
+```{code-cell}
+@jax.jit
+def eval_step(params, batch):
+  logits = CNN(training=False).apply({'params': params}, batch['image'])
```
Comment (cgarciae) [inline docs/tutorials/dropout_tutorial.md:159]: Same situation but `train=False`.

**Comment by marcvanzee** [comment]
> (removed myself as a reviewer because I don't have time to properly review this ... maybe @marcvanzee can have a look instead? )

Sure, added myself as a reviewer.

**Comment by marcvanzee** [comment]
Thanks for adding this @8bitmp3!

I have some questions / remarks:
* It seems RTD does not build due to a time out, which is because you are doing a full training run in the tutorial. The code blocks are ran as a part of our CI on CPU, and we try to keep them short.

* More generally, I am a bit worried that our documentation for Dropout is getting somewhat unwieldy now:
    * This guides explains Dropout in an end-to-end example
    * The "sharp bits" also explains Dropout in detail (but still contains some mistakes)

I was actually thinking that we would create a single "Using Dropout" guide, similar to @cgarciae 's ["Using BatchNorm"](https://flax--2536.org.readthedocs.build/en/2536/guides/batch_norm.html). In this guide we would show how one goes about adding Dropout to their model, with side-by-side diffs showing which parts have to be changed.

So therefore, before we start writing any new documentation for Dropout, I think we should replace the "sharp bits" content with a guide where we explain how to do Dropout. Once we are happy with that we can start thinking of more content for Dropout.

It is of course great you are writing these end-to-end tutorials proactively, similar to the one you are proposing in [Variational autoencoder with a learning rate schedule](https://github.com/google/flax/pull/2650/), but I suggest you first discuss what content you are planning to put in these tutorials, otherwise you might loose a lot of time writing them if it turns out we want to have other content.

Anyway, let's chat more offline!

File: docs/tutorials/dropout_tutorial.md (hunk 1 above)
Comment (8bitmp3) [reply docs/tutorials/dropout_tutorial.md:80]: Thanks @cgarciae 👍 

Just to confirm, do you mean?

```
  def __call__(self, x, train = True):
```

Comment (cgarciae) [reply docs/tutorials/dropout_tutorial.md:80]: Yes! Usually just `train: bool` with no default value to be very explicit.

Comment (cgarciae) [reply docs/tutorials/dropout_tutorial.md:80]: Yes! Usually just `train: bool` with no default value to be very explicit.

**Comment by 8bitmp3** [comment]
> The "sharp bits" also explains Dropout in detail (but still contains some mistakes)

Reviewed by @levskaya but we can revise it again.

> This guides explains Dropout in an end-to-end example

It's an end-to-end tutorial that uses dropout with a particular task in mind, like classification.

> I was actually thinking that we would create a single "Using Dropout" guide,

Let's do this! 👍 

> suggest you first discuss what content you are planning to put in these tutorials, otherwise you might loose a lot of time writing them if it turns out we want to have other content.

This doc was presented to the team 2 weeks ago.

**Comment by 8bitmp3** [comment]
Publishing a dropout _guide_, closing this PR. Discussed with marcvanzee 👍
